internal/sqldb/sqlquery/observation.go

```diff
@@ -0,0 +1,170 @@
+package sqlquery
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+
+	pb "github.com/datacommonsorg/mixer/internal/proto"
+	pbv2 "github.com/datacommonsorg/mixer/internal/proto/v2"
+	"github.com/datacommonsorg/mixer/internal/server/v2/shared"
+	"github.com/datacommonsorg/mixer/internal/util"
+	"google.golang.org/protobuf/proto"
+)
+
+// GetObservations fetches observations from the specified SQL database.
+func GetObservations(
+	ctx context.Context,
+	sqlClient *sql.DB,
+	sqlProvenances map[string]*pb.Facet,
+	variables []string,
+	entities []string,
+	queryDate string,
+	filter *pbv2.FacetFilter,
+) (*pbv2.ObservationResponse, error) {
+	if sqlClient == nil || len(variables) == 0 {
+		return newObservationResponse(variables), nil
+	}
+
+	// Query SQL.
+	rows, err := sqlClient.Query(getObservationsSQLQuery(variables, entities, queryDate))
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	// Generate intermediate response.
+	intermediateResponse, err := generateIntermediateResponse(rows, sqlProvenances)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate ObservationResponse.
+	return generateObservationResponse(intermediateResponse, variables, queryDate), nil
+}
+
+func generateObservationResponse(intermdiate *intermediateObservationResponse, variables []string, queryDate string) *pbv2.ObservationResponse {
+	response := newObservationResponse(variables)
+
+	for byFacetKey, byFacetValue := range intermdiate.byFacet {
+		variable, entity, facetId := byFacetKey.variable, byFacetKey.entity, byFacetKey.facetId
+		observations, facet := byFacetValue.observations, byFacetValue.facet
+		if response.ByVariable[variable].ByEntity[entity] == nil {
+			response.ByVariable[variable].ByEntity[entity] = &pbv2.EntityObservation{
+				OrderedFacets: []*pbv2.FacetObservation{},
+			}
+		}
+		if queryDate == shared.LATEST {
+			observations = observations[len(observations)-1:]
+		}
+		response.ByVariable[variable].ByEntity[entity].OrderedFacets = append(
+			response.ByVariable[variable].ByEntity[entity].OrderedFacets,
+			&pbv2.FacetObservation{
+				FacetId:      facetId,
+				Observations: observations,
+				ObsCount:     int32(len(observations)),
+				EarliestDate: observations[0].Date,
+				LatestDate:   observations[len(observations)-1].Date,
+			},
+		)
+		response.Facets[facetId] = facet
+	}
+
+	return response
+}
+
+func generateIntermediateResponse(rows *sql.Rows, cachedProvenances map[string]*pb.Facet) (*intermediateObservationResponse, error) {
+	defer rows.Close()
```

Comment:
I find the defer odd here...meaning this function can not be used standalone.. maybe just merge the two functions as they are internal anyway

Reply:
This was an oversight - thanks for catching. I'm already closing the rows in `GetObservations` so not needed here.